server/cluster/unsafe_recovery_controller.go

```diff
@@ -120,462 +184,747 @@ func (u *unsafeRecoveryController) RemoveFailedStores(failedStores map[uint64]st
+}
+
+func newRegionTree() *regionTree {
+	return &regionTree{
+		regions: make(map[uint64]*regionItem),
+		tree:    btree.New(defaultBTreeDegree),
 	}
-	region.Peers = newPeerList
 }
 
-type peerStorePair struct {
-	peer    *pdpb.PeerReport
-	storeID uint64
+func (t *regionTree) contains(regionID uint64) bool {
+	_, ok := t.regions[regionID]
+	return ok
 }
 
-func getOverlapRanges(tree *btree.BTree, region *metapb.Region) []*metapb.Region {
-	var overlapRanges []*metapb.Region
-	tree.DescendLessOrEqual(regionItem{region}, func(item btree.Item) bool {
-		if bytes.Compare(item.(regionItem).region.StartKey, region.StartKey) < 0 && bytes.Compare(item.(regionItem).region.EndKey, region.StartKey) > 0 {
-			overlapRanges = append(overlapRanges, item.(regionItem).region)
-		}
-		return false
-	})
+// getOverlaps gets the regions which are overlapped with the specified region range.
+func (t *regionTree) getOverlaps(item *regionItem) []*regionItem {
+	// note that find() gets the last item that is less or equal than the region.
+	// in the case: |_______a_______|_____b_____|___c___|
+	// new region is     |______d______|
+	// find() will return regionItem of region_a
+	// and both startKey of region_a and region_b are less than endKey of region_d,
+	// thus they are regarded as overlapped regions.
+	result := t.find(item)
+	if result == nil {
+		result = item
+	}
 
-	tree.AscendGreaterOrEqual(regionItem{region}, func(item btree.Item) bool {
-		if len(region.EndKey) != 0 && bytes.Compare(item.(regionItem).region.StartKey, region.EndKey) > 0 {
+	end := item.Region().GetEndKey()
+	var overlaps []*regionItem
+	t.tree.AscendGreaterOrEqual(result, func(i btree.Item) bool {
+		over := i.(*regionItem)
+		if len(end) > 0 && bytes.Compare(end, over.Region().GetStartKey()) <= 0 {
 			return false
 		}
-		overlapRanges = append(overlapRanges, item.(regionItem).region)
+		overlaps = append(overlaps, over)
 		return true
 	})
-	return overlapRanges
+	return overlaps
 }
 
-func (u *unsafeRecoveryController) generateRecoveryPlan() {
-	u.Lock()
-	defer u.Unlock()
-	newestRegionReports := make(map[uint64]*pdpb.PeerReport)
-	var allPeerReports []*peerStorePair
+// find is a helper function to find an item that contains the regions start key.
+func (t *regionTree) find(item *regionItem) *regionItem {
+	var result *regionItem
+	t.tree.DescendLessOrEqual(item, func(i btree.Item) bool {
+		result = i.(*regionItem)
+		return false
+	})
+
+	if result == nil || !result.Contains(item.Region().GetStartKey()) {
+		return nil
+	}
+
+	return result
+}
+
+// Insert the peer report of one region int the tree.
+// It finds and deletes all the overlapped regions first, and then
+// insert the new region.
+func (t *regionTree) insert(item *regionItem) bool {
+	overlaps := t.getOverlaps(item)
+
+	if t.contains(item.Region().GetId()) {
+		// it's ensured by the `buildUpFromReports` that only insert the latest peer of one region.
+		panic("region shouldn't be updated twice")
```

Comment:
It's better to remove all panic since if it failed to recover, we shouldn't let the whole service down.